sound/soc/sof/pcm.c

```diff
@@ -318,6 +318,19 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 	snd_sof_pcm_platform_trigger(sdev, substream, cmd);
 
 	/* send IPC to the DSP */
+	/*
+	 * The trigger action is in non-atomic environment, which
+	 * is set in topology parsing. Here is the details.
+	 * In soc_new_pcm(), as only FE PCM will really create pcm
+	 * (BE no_pcm is 1), so the pcm->nonatomic will be set to the
+	 * FE link's nonatomic in ASoC:
+	 * pcm->nonatomic = rtd->dai_link->nonatomic
+	 * when playback/capture, function __snd_pcm_stream_lock_mode()
+	 * will check whether pcm nonatomic is set or not. If the pcm is
+	 * set to nonaotmic, alsa will set the environment as nonatomic.
+	 * So as SOF FE has already set it as nonatomic in the topology,
+	 * calling sof_ipc_tx_message() is safe.
```

Comment:
@libinyang this is too confusing to read. I think it is already quite clear that the trigger is in the non-atomic context. The information we need to clarify is:

The SOF driver sets nonatomic to true for both BE and FE dai links. 
But the legacy driver only sets nonatomic for FE dai links and not for BE dai links. Why is this different?

Reply:
@libinyang I think it is meaningless to set BE dai link to nonatomic, which is what we do now. I think we should change that and then clarify here that FE dai links are set to non-atomic because the trigger is always in non-atomic context.

Reply:
> @libinyang I think it is meaningless to set BE dai link to nonatomic, which is what we do now. I think we should change that and then clarify here that FE dai links are set to non-atomic because the trigger is always in non-atomic context.

@ranj063 
Yes, you got it :) 
Do we set nonatomic for BE? I didn't find we define nonatomic for BE. Anyway, it is OK for us to define nonatomic for BE because it will be ignored whatever it is set.